crates/shared/src/price_estimation/rate_limited.rs

```diff
@@ -0,0 +1,162 @@
+use crate::{
+    price_estimation::{single_estimate, PriceEstimating, PriceEstimationError, Query},
+    rate_limiter::RateLimiter,
+};
+use futures::stream::{BoxStream, StreamExt};
+use std::{
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+pub trait RateLimitedPriceEstimatorExt {
+    /// Wraps an existing price estimator in rate limiting one.
+    fn rate_limited(
+        self,
+        rate_limiter: RateLimiter,
+        healthy_response_time: Duration,
+    ) -> RateLimitedPriceEstimator;
+}
+
+impl<T: PriceEstimating + 'static> RateLimitedPriceEstimatorExt for T {
+    fn rate_limited(
+        self,
+        rate_limiter: RateLimiter,
+        healthy_response_time: Duration,
+    ) -> RateLimitedPriceEstimator {
+        RateLimitedPriceEstimator {
+            inner: Arc::new(self),
+            rate_limiter,
+            healthy_response_time,
+        }
+    }
+}
+
+/// A price estimator which backs off when it detects long response times.
+/// Executes a list of queries sequentially.
+pub struct RateLimitedPriceEstimator {
+    inner: Arc<dyn PriceEstimating>,
+    rate_limiter: RateLimiter,
+    healthy_response_time: Duration,
+}
+
+impl RateLimitedPriceEstimator {
+    /// Wraps an existing price estimator in rate limiting one.
+    pub fn new(
+        inner: Arc<dyn PriceEstimating>,
+        rate_limiter: RateLimiter,
+        healthy_response_time: Duration,
+    ) -> Self {
+        Self {
+            inner,
+            rate_limiter,
+            healthy_response_time,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl PriceEstimating for RateLimitedPriceEstimator {
+    fn estimates<'a>(
+        &'a self,
+        queries: &'a [Query],
+    ) -> BoxStream<'_, (usize, super::PriceEstimateResult)> {
+        let start = Instant::now();
+        futures::stream::iter(queries.iter())
+            .enumerate()
+            .then(move |(index, query)| async move {
+                let result = self
+                    .rate_limiter
+                    .execute(single_estimate(&*self.inner, query), |_| {
+                        let elapsed = start.elapsed();
+                        elapsed > self.healthy_response_time
+                    })
+                    .await
+                    .map_err(PriceEstimationError::from);
+
+                (index, result)
+            })
```

Comment:
Another observation in favor of this:
Imagine our healthy time is 10 seconds and we want 10 estimates. Then if we call an estimator one by one it would be allowed to return each of those results after 10 seconds for a total of 100 seconds and be considered healthy. But if we called in a batching way then only the first response would be healthy because for the remaining 9 with the current code they are considered out of time.

Reply:
Those are good points.
I decided to go for the option of rate limiting every estimator except `Baseline`.

Reply:
I still think it would be cleaner to integrate with "single price estimators" directly rather than keeping this code. I am also fine with doing that later.

Reply:
I looked into how that could be done and I believe the error propagation of `RateLimiter` needs to be improved in order to be compatible with that idea.
On the other hand the way the error propagation happens right now is not ideal anyway so I might as well fix that and then integrate the `RateLimiter` into each `PriceEstimator`.